pkg/retagger/job_pattern.go

```diff
@@ -0,0 +1,130 @@
+
+	// Create a reference to the external registry.
+	o := dockerRegistry.Options{
+		Logf:          dockerRegistry.Quiet,
+		DoInitialPing: false,
+	}
+	externalRegistry, err := dockerRegistry.NewCustom(fmt.Sprintf("https://%s", job.Source.RepoPath), o)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Find tags which match the pattern.
+	matches, err := getExternalTagMatches(externalRegistry, job.Source.FullImageName, job.SourcePattern)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	if len(matches) == 0 {
+		r.logger.Log("level", "warn", "message", fmt.Sprintf("No upstream image tags were found matching the pattern %s", job.SourcePattern))
+	} else {
+		r.logger.Log("level", "debug", "message", fmt.Sprintf("Found %d upstream tags which match the pattern %s", len(matches), job.SourcePattern))
+	}
+
+	// Get SHA/Tag pairs from our quay registry.
+	quayTagMap, err := r.GetTagDetails(job.Destination.Image)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	jobs := []SingleJob{}
+
+	for _, match := range matches {
+
+		_, exists := quayTagMap[match]
+
+		if !exists {
+			// Tag is new - get SHA and tag it.
+			newDigest, err := externalRegistry.ManifestDigest(job.Source.FullImageName, match)
+			if err != nil {
+				return nil, microerror.Mask(err)
+			}
+			sourceSHA := strings.TrimPrefix(newDigest.String(), "sha256:")
+			// Create job with new SHA.
+			j := SingleJob{
+
+				Source: job.Source,
+
+				Options: job.Options,
+			}
+			// Override Source options from our pattern.
+			j.Source.Tag = match
+			j.Source.SHA = sourceSHA
+			j.Destination = GetDestinationForJob(&j, r)
+			jobs = append(jobs, j)
+		}
+	}
+
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("Compiled %d jobs to process", len(jobs)))
+
+	return jobs, nil
+}
+
+func (job *PatternJob) GetOptions() JobOptions {
+	return job.Options
+}
+
+func (job *PatternJob) GetSource() Source {
+	return job.Source
+}
+
+// PatternJobFromJobDefinition converts a JobDefinition into a PatternJob.
+func PatternJobFromJobDefinition(jobDef *JobDefinition, r *Retagger) *PatternJob {
+	job := &PatternJob{
+		SourcePattern: jobDef.SourcePattern,
+		Source:        GetSourceForJob(jobDef, r),
+
+		Options: jobDef.Options,
+	}
+	job.Destination = GetDestinationForJob(job, r)
+	return job
+}
+
+// getExternalTagMatches searches the given docker registry for tags matching the given pattern.
+func getExternalTagMatches(r *dockerRegistry.Registry, image string, pattern string) ([]string, error) {
+	// Make sure our pattern is valid.
+	regex, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Get the tags for this image from the external registry.
+	externalRegistryTags, err := r.Tags(image)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	// Find tags matching our configured pattern.
+	matches := []string{}
```

Comment:
I would prefer `var matches []string` here. Go allows you to append to a `nil` slice and it avoid unnecessary initialization.

Reply:
I like that much better. Changed in a couple other places as well